cmd/snap-seccomp/main_test.go

```diff
@@ -0,0 +1,404 @@
+)
+
+// Hook up check.v1 into the "go test" runner
+func Test(t *testing.T) { TestingT(t) }
+
+type snapSeccompSuite struct{}
+
+var _ = Suite(&snapSeccompSuite{})
+
+func goArchToScmpArch(goarch string) uint32 {
+	switch goarch {
+	case "386":
+		return main.ScmpArchX86
+	case "amd64":
+		return main.ScmpArchX86_64
+	case "arm":
+		return main.ScmpArchARM
+	case "arm64":
+		return main.ScmpArchAARCH64
+	case "ppc64le":
+		return main.ScmpArchPPC64LE
+	case "s390x":
+		return main.ScmpArchS390X
+	case "ppc":
+		return main.ScmpArchPPC
+	}
+	panic(fmt.Sprintf("cannot map goarch %q to a seccomp arch", goarch))
+}
+
+func decodeBpfFromFile(p string) ([]bpf.Instruction, error) {
+	var ops []bpf.Instruction
+	var rawOp bpf.RawInstruction
+
+	r, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	for {
+		err = binary.Read(r, nativeEndian(), &rawOp)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, rawOp.Disassemble())
+	}
+
+	return ops, nil
+}
+
+func parseBpfInput(s string) (*main.SeccompData /* *seccompData */, error) {
+	// syscall;arch;arg1,arg2...
+	l := strings.Split(s, ";")
+
+	sc, err := seccomp.GetSyscallFromName(l[0])
+	if err != nil {
+		return nil, err
+	}
+
+	var arch uint32
+	if len(l) < 2 || l[1] == "native" {
+		arch = goArchToScmpArch(runtime.GOARCH)
+	} else {
+		arch = goArchToScmpArch(l[1])
+	}
+
+	var syscallArgs [6]uint64
+	if len(l) > 2 {
+		args := strings.Split(l[2], ",")
+		for i := range args {
+			if nr, err := strconv.ParseUint(args[i], 10, 64); err == nil {
+				syscallArgs[i] = nr
+			} else {
+				syscallArgs[i] = main.SeccompResolver[args[i]]
+			}
+		}
+	}
+	sd := &main.SeccompData{}
+	sd.SetArch(arch)
+	sd.SetNr(sc)
+	sd.SetArgs(syscallArgs)
+	return sd, nil
+}
+
+// Endianness detection.
+func nativeEndian() binary.ByteOrder {
+	// Credit matt kane, taken from his gosndfile project.
+	// https://groups.google.com/forum/#!msg/golang-nuts/3GEzwKfRRQw/D1bMbFP-ClAJ
+	// https://github.com/mkb218/gosndfile
+	var i int32 = 0x01020304
+	u := unsafe.Pointer(&i)
+	pb := (*byte)(u)
+	b := *pb
+	isLittleEndian := b == 0x04
+
+	if isLittleEndian {
+		return binary.LittleEndian
+	} else {
+		return binary.BigEndian
+	}
+}
+
+func simulateBpf(c *C, seccompWhitelist, bpfInput string, expected int) {
+	outPath := filepath.Join(c.MkDir(), "bpf")
+	err := main.Compile([]byte(seccompWhitelist), outPath)
+	c.Assert(err, IsNil)
+
+	ops, err := decodeBpfFromFile(outPath)
+	c.Assert(err, IsNil)
+
+	vm, err := bpf.NewVM(ops)
+	c.Assert(err, IsNil)
+
+	bpfSeccompInput, err := parseBpfInput(bpfInput)
+	c.Assert(err, IsNil)
+
+	buf2 := (*[64]byte)(unsafe.Pointer(bpfSeccompInput))
+
+	out, err := vm.Run(buf2[:])
+	c.Assert(err, IsNil)
+	c.Check(out, Equals, expected, Commentf("unexpected result for %q (input %q), got %v expected %v", seccompWhitelist, bpfInput, out, expected))
+}
+
+func (s *snapSeccompSuite) TestCompile(c *C) {
```

Comment:
Thanks! I added the following:
```
// TestCompile will test the input from our textual seccomp whitelist
// against a kernel syscall input that may contain arguments. The test
// is performed by running the compiled bpf program on a virtual bpf
// machine. Each test needs to declare what output from the VM it expects.
//
// This output is usually main.SeccompRet{Allow,Kill}. It is recommended
// to test the allow and kill case for each new syscall and each argument.
```
I hope this makes it a bit clearer. I can also change the test and split into one allow and one kill test. However to me it feels cleaner when the too are close(ish) together. But I'm happy to change if you/gustavo prefer a different style.